File: jetbrains/src/test/kotlin/com/almasix/ide/CallSiteDetectorTest.kt
package com.almasix.ide

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class CallSiteDetectorTest {
    @Test
    fun routeCall() {
        val site = CallSiteDetector.detect("""route("hom""")
        assertNotNull(site)
        assertEquals(SymbolKind.ROUTE, site!!.kind)
        assertEquals("hom", site.prefix)
    }

    @Test
    fun viewCall() {
        val site = CallSiteDetector.detect("""view('welcome""")
        assertEquals(SymbolKind.VIEW, site!!.kind)
        assertEquals("welcome", site.prefix)
    }

    @Test
    fun configCall() {
        val site = CallSiteDetector.detect("""config("app.na""")
        assertEquals(SymbolKind.CONFIG, site!!.kind)
        assertEquals("app.na", site.prefix)
    }

    @Test
    fun gateCan() {
        val site = CallSiteDetector.detect("""can("edit-""")
        assertEquals(SymbolKind.GATE, site!!.kind)
    }

    @Test
    fun withRelation() {
        val site = CallSiteDetector.detect("""Post.with_("comm""")
        assertEquals(SymbolKind.RELATION, site!!.kind)
        assertEquals("comm", site.prefix)
    }

    @Test
    fun validationRule() {
        val site = CallSiteDetector.detect("""validate({"email": "requ""")
        assertEquals(SymbolKind.VALIDATION, site!!.kind)
        assertEquals("requ", site.prefix)
    }

    @Test
    fun prismInclude() {
        val site = CallSiteDetector.detect("""@include('layouts.""")
        assertEquals(SymbolKind.VIEW, site!!.kind)
    }

    @Test
    fun prismComponent() {
        val site = CallSiteDetector.detect("""<x-alert.""")
        assertEquals(SymbolKind.COMPONENT, site!!.kind)
        assertEquals("alert.", site.prefix)
    }

    @Test
    fun prismDirective() {
        val site = CallSiteDetector.detect("""  @en""")
        assertEquals(SymbolKind.DIRECTIVE, site!!.kind)
        assertEquals("en", site.prefix)
    }

    @Test
    fun envCall() {
        val site = CallSiteDetector.detect("""env("APP_""")
        assertEquals(SymbolKind.ENV, site!!.kind)
    }
}

class AlmasixIndexParseTest {
    @Test
    fun parsesFixture() {
        val json = """
            {
              "base_path": "/tmp/app",
              "ok": true,
              "error": null,
              "views": {"welcome": "/tmp/app/resources/views/welcome.prism.html"},
              "routes": {
                "home": {"name": "home", "uri": "/", "methods": ["GET"], "path": null, "line": 0}
              },
              "config_keys": ["app.name", "app.env"],
              "translation_keys": ["messages.hello"],
              "middleware_aliases": ["web", "auth"],
              "env_keys": {"APP_KEY": {"name": "APP_KEY"}},
              "tables": {
                "users": {
                  "name": "users",
                  "columns": {"id": {"name": "id"}, "email": {"name": "email"}},
                  "detail": "users"
                }
              },
              "model_metadata": {
                "User": {
                  "module": "user",
                  "fillable": ["email"],
                  "casts": {"id": "int"},
                  "relations": ["posts"]
                }
              },
              "relations": {"User": ["posts"]},
              "casts": ["int", "datetime"],
              "components": {"alert": "/tmp/x"},
              "gates": ["update"],
              "disks": ["local"],
              "queues": ["sync"],
              "caches": ["file"],
              "mailers": ["smtp"],
              "inertia_pages": ["Dashboard"],
              "smith_commands": ["serve", "ide:index"],
              "validation_rules": ["required", "email"],
              "directives": ["if", "endif"],
              "view_helpers": [{"name": "auth"}],
              "view_shared": {},
              "view_data": {"welcome": {"title": {"name": "title"}}},
              "vite_entries": {"resources/js/app.js": "/tmp/app.js"},
              "controller_actions": {"WelcomeController": ["index"]}
            }
        """.trimIndent()
        val index = AlmasixIndexLoader.parse(json)
        assertTrue(index.ok)
        assertTrue(index.views.contains("welcome"))
        assertEquals("/", index.routes["home"]!!.uri)
        assertTrue(index.configKeys.contains("app.name"))
        assertTrue(index.tables["users"]!!.columns.contains("email"))
        assertEquals(listOf("posts"), index.relations["User"])
        assertTrue(index.validationRules.contains("required"))
        assertTrue(index.known(SymbolKind.ROUTE, "home"))
        assertTrue(!index.known(SymbolKind.ROUTE, "missing"))

        val site = CallSiteDetector.Site(SymbolKind.ROUTE, "ho")
        val items = AlmasixCompletionContributor.symbolsFor(index, site)
        assertTrue(items.any { it.first == "home" })
    }
}

File: vscode/src/core/__tests__/callSiteDetector.test.ts
import { describe, expect, it } from "vitest";
import { CallSiteDetector } from "../callSiteDetector";
import { CompletionCatalog } from "../completionCatalog";
import { ModelResolver } from "../modelResolver";
import { AlmasixIndex, SymbolKind } from "../types";

describe("CallSiteDetector", () => {
  it("detects route call", () => {
    const site = CallSiteDetector.detect(`route("hom`);
    expect(site).not.toBeNull();
    expect(site!.kind).toBe(SymbolKind.ROUTE);
    expect(site!.prefix).toBe("hom");
  });

  it("detects view call", () => {
    const site = CallSiteDetector.detect(`view('welcome`);
    expect(site!.kind).toBe(SymbolKind.VIEW);
    expect(site!.prefix).toBe("welcome");
  });

  it("detects config and env", () => {
    expect(CallSiteDetector.detect(`config("app.na`)!.kind).toBe(SymbolKind.CONFIG);
    expect(CallSiteDetector.detect(`env("APP_`)!.kind).toBe(SymbolKind.ENV);
  });

  it("detects dotenv bare key", () => {
    const site = CallSiteDetector.detect("QUEUE_CON", true);
    expect(site!.kind).toBe(SymbolKind.ENV);
    expect(site!.prefix).toBe("QUEUE_CON");
  });

  it("detects component x- tags", () => {
    const site = CallSiteDetector.detect(`<x-alert.`);
    expect(site!.kind).toBe(SymbolKind.COMPONENT);
    expect(site!.prefix).toBe("alert.");
  });

  it("detects validation rules", () => {
    const site = CallSiteDetector.detect(`validate({"email": "requ`);
    expect(site!.kind).toBe(SymbolKind.VALIDATION);
    expect(site!.prefix).toBe("requ");
  });

  it("detects casts key and value", () => {
    expect(CallSiteDetector.detect(`casts = {"ema`)!.kind).toBe(SymbolKind.MODEL_ATTR);
    expect(CallSiteDetector.detect(`casts = {"email": "dat`)!.kind).toBe(SymbolKind.CAST);
  });

  it("detects fillable list keys", () => {
    const modelSrc = `
class User(Model):
    fillable = ["ema
`.trim();
    expect(CallSiteDetector.detect(modelSrc)!.kind).toBe(SymbolKind.MODEL_ATTR);
  });

  it("detects Author.query().where chain", () => {
    const chain = CallSiteDetector.detect(`Author.query().where("ema`);
    expect(chain!.kind).toBe(SymbolKind.COLUMN);
    expect(chain!.receiver).toBe("Author");
    expect(chain!.prefix).toBe("ema");
  });

  it("detects Author.create(email= kwarg", () => {
    const kw = CallSiteDetector.detect(`Author.create(email=`);
    expect(kw!.kind).toBe(SymbolKind.COLUMN);
    expect(kw!.receiver).toBe("Author");
    expect(kw!.prefix).toBe("email");
  });

  it("detects DICT create column keys", () => {
    const site = CallSiteDetector.detect(`Author.create({"ema`);
    expect(site!.kind).toBe(SymbolKind.COLUMN);
    expect(site!.receiver).toBe("Author");
    expect(site!.prefix).toBe("ema");
  });

  it("detects chained redirect().route", () => {
    expect(CallSiteDetector.detect(`redirect().route("hom`)!.kind).toBe(SymbolKind.ROUTE);
  });

  it("detects @if directive prefix", () => {
    const site = CallSiteDetector.detect(`  @if`);
    expect(site!.kind).toBe(SymbolKind.DIRECTIVE);
    expect(site!.prefix).toBe("if");
  });

  it("does not dump columns for table. attr", () => {
    const tableAttr = CallSiteDetector.detect("table.ema");
    expect(tableAttr!.kind).toBe(SymbolKind.ATTR);
    const items = CompletionCatalog.symbolsFor(
      new AlmasixIndex({
        ok: true,
        tables: {
          users: { columns: { email: {} } },
        },
      }),
      tableAttr!,
    );
    expect(items).toEqual([]);
  });

  it("detects auth().user() attributes", () => {
    const auth = CallSiteDetector.detect("auth().user().ema");
    expect(auth!.kind).toBe(SymbolKind.ATTR);
    expect(auth!.receiver).toBe(ModelResolver.AUTH_USER_SENTINEL);
  });
});
